docs/decisions/0054-python-structured-output.md

```diff
@@ -0,0 +1,312 @@
+
+With the introduction of Structured Outputs, OpenAI models are now able to strictly adhere to developer-provided JSON schemas. This feature eliminates the need for cumbersome workarounds and provides a more streamlined, efficient way to ensure consistency and reliability in model outputs. Integrating Structured Outputs into the Semantic Kernel orchestration SDK will enable developers to create more powerful, schema-compliant applications, reduce errors, and improve overall productivity.
+
+## Using Structured Outputs
+
+### 1. Function Calling
+
+Structured outputs can be used for function calling to ensure adherence to a schema. In Python, using OpenAI's latest SDK, developers can use the `pydantic_function_tool()` method to convert a Pydantic model into the required `structured output` function-calling JSON schema:
+
+
+```python
+client = AsyncOpenAI()
+
+# Note that the SDK method is `client.beta.chat.completions.parse` instead of our current `client.chat.completions.create` call.
+completion = await client.beta.chat.completions.parse(
+    model="gpt-4o-2024-08-06",
+    messages=[
+        {
+            "role": "system",
+            "content": "You are a helpful assistant. The current date is August 6, 2024. You help users query for the data they are looking for by calling the query function.",
+        },
+        {
+            "role": "user",
+            "content": "look up all my orders in may of last year that were fulfilled but not delivered on time",
+        },
+    ],
+    tools=[
+        openai.pydantic_function_tool(SomeClass), # This is a new method provided by OpenAI to handle Pydantic models
+    ],
+)
+
+print(completion.choices[0].message)
+```
+
+This produces the following JSON Schema:
+
+```json
+{
+  "type": "function",
+  "function": {
+    "name": "math-add_numbers",
+    "description": "Adds two numbers together and provides the result",
+    "strict": true, // new key-value pair for structured output function calling
+    "parameters": {
+      "type": "object",
+      "properties": {
+        "number_one": {
+          "type": "integer",
+          "description": "The first number to add"
+        },
+        "number_two": {
+          "type": "integer",
+          "description": "The second number to add"
+        }
+      },
+      "required": ["number_one", "number_two"],
+      "additionalProperties": false // new key-value pair for structured output function calling
+    }
+  }
+}
+```
+
+For non-Pydantic models, SK will need to manually add the `strict` and `additionalProperties` key-value pairs to the JSON schema, which is currently built as part of the `KernelParameterMetadata`.
```

Comment:
Add verbiage about the necessity of doing this: we need to do this because the model provided isn't Pydantic, and we'd be unable to use the provided library method (`openai.pydantic_function_tool`)

Reply:
Irrelevant for this ADR - will be covered in a new ADR. Closing.